lib/std/optionals.nim

```diff
@@ -0,0 +1,331 @@
+#
+#
+#            Nim's Runtime Library
+#        (c) Copyright 2022 Nim Contributors
+#
+#    See the file "copying.txt", included in this
+#    distribution, for details about the copyright.
+#
+
+##[
+This module implements a type which encapsulate an optional value.
+
+A value of type `Optional[T]` either contains a value `x` (represented as
+`some(x)`) or is empty (`none(T)`).
+
+This can be useful when you have a value that can be present or not. The
+absence of a value is often represented by `nil`, but that is not always
+available, nor is it always a good solution.
+
+
+Basic usage
+===========
+
+Let's start with an example: a procedure that finds the index of a character
+in a string.
+]##
+
+runnableExamples:
+  proc find(haystack: string, needle: char): Optional[int] =
+    for i, c in haystack:
+      if c == needle:
+        return some(i)
+    return none(int)  # This line is actually optional,
+                      # because the default is empty
+
+  let found = "abc".find('c')
+  assert found.isSome and found.get() == 2
+
+##[
+The `get` operation demonstrated above returns the underlying value, or
+raises `UnpackDefect` if there is no value. Note that `UnpackDefect`
+inherits from `system.Defect` and should therefore never be caught.
+Instead, rely on checking if the optional contains a value with the
+`isSome <#isSome,Optional[T]>`_ and `isNone <#isNone,Optional[T]>`_ procs.
+]##
+
+
+import typetraits
+
+when defined(nimPreviewSlimSystem):
+  import std/assertions
+
+
+type
+  SomePointer = ref | ptr | pointer | proc # xxx what about cstring? proc vs closure?
+  Optional*[T] = object
+    ## An optional type that may or may not contain a value of type `T`.
+    val: T
+    has: bool
+
+  UnpackDefect* = object of Defect
+
+template maybeOptional*[T](_: typedesc[T]): untyped =
+  ## Return `T` if T is `ref | ptr | pointer | proc`, else `Optional[T]`
+  runnableExamples:
+    assert maybeOptional(ref int) is ref int
+    assert maybeOptional(int) is Optional[int]
+  when T is SomePointer: T
+  else: Optional[T]
+
+proc optional*[T](val: sink T): Optional[T] {.inline.} =
+  ## Can be used to convert a pointer type (`ptr`, `pointer`, `ref` or `proc`) to an optional type.
+  ## It converts `nil` to `none(T)`. When `T` is no pointer type, this is equivalent to `some(val)`.
+  ##
+  ## **See also:**
+  ## * `some proc <#some,T>`_
+  ## * `none proc <#none,typedesc>`_
+  runnableExamples:
+    type
+      Foo = ref object
+    assert optional[Foo](nil).isNone
+    assert optional(Foo(nil)).isNone
+    assert some(Foo(nil)).isSome
+    assert optional(42).isSome
+  when T is SomePointer:
+    result.val = val
+    result.has = val != nil
+  else:
+    result.val = val
+    result.has = true
+
+proc some*[T](val: sink T): Optional[T] {.inline.} =
```

Comment:
```suggestion
proc some*[T](O: type Optional, val: sink T): Optional[T] {.inline.} =
```

the above signature will make the syntax `Optional.some(value)` which may be an advantage when wanting to avoid name clashes